File: solutions/java/src/main/java/com/epi/HouseMajority.java
package com.epi;

import java.util.Arrays;
import java.util.Random;

import static java.lang.Math.ceil;

public class HouseMajority {
  // @include
  public static double houseMajority(double[] prob, int n) {
    // Initializes DP table.
    double[][] P = new double[n + 1][n + 1];
    for (double[] element : P) {
      Arrays.fill(element, -1.0);
    }

    // Accumulates the probabilities of majority cases.
    double probSum = 0.0;
    for (int r = (int)ceil(0.5 * n); r <= n; ++r) {
      probSum += houseMajorityHelper(prob, r, n, P);
    }
    return probSum;
  }

  // prob is the probability that each Republican wins.
  // r is the number of Republicans wins, and n is the number of elections.
  private static double houseMajorityHelper(double[] prob, int r, int n,
                                            double[][] P) {
    if (r > n) {
      return 0.0; // Base case: not enough Republicans.
    } else if (r == 0 && n == 0) {
      return 1.0; // Base case.
    } else if (r < 0) {
      return 0.0;
    }

    if (P[r][n] == -1.0) {
      P[r][n] = houseMajorityHelper(prob, r - 1, n - 1, P) * prob[n - 1] +
                houseMajorityHelper(prob, r, n - 1, P) * (1.0 - prob[n - 1]);
    }
    return P[r][n];
  }
  // @exclude

  private static void printVector(double[] prob) {
    Arrays.sort(prob);
    for (int i = 0; i < prob.length; ++i) {
      System.out.println(String.format("%s:%s:%s", i, 100 * prob[i],
                                       ((i + 1) % 10 == 0) ? "\n" : " "));
    }
    System.out.println();
  }

  public static void main(String[] args) {
    Random gen = new Random();
    int n = 435;
    double[] prob = new double[n + 1];
    prob[0] = 0.0;
    for (int i = 1; i < n + 1; ++i) {
      prob[i] = gen.nextDouble();
    }
    printVector(prob);
    double ans = houseMajority(prob, n);
    assert 0.0 <= ans&& ans <= 1.0;
    System.out.println();
  }
}

File: static/solutions/House_majority.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::uniform_real_distribution;
using std::vector;

double HouseMajorityHelper(const vector<double>&, int, int,
                           vector<vector<double>>*);

// @include
double HouseMajority(const vector<double>& prob, int n) {
  // Initializes DP table.
  vector<vector<double>> P(n + 1, vector<double>(n + 1, -1.0));

  // Accumulates the probabilities of majority cases.
  double prob_sum = 0.0;
  for (int r = ceil(0.5 * n); r <= n; ++r) {
    prob_sum += HouseMajorityHelper(prob, r, n, &P);
  }
  return prob_sum;
}

// prob is the probability that each Republican wins.
// r is the number of Republicans wins, and n is the number of elections.
double HouseMajorityHelper(const vector<double>& prob, int r, int n,
                           vector<vector<double>>* P_ptr) {
  if (r > n) {
    return 0.0;  // Base case: not enough Republicans.
  } else if (r == 0 && n == 0) {
    return 1.0;  // Base case.
  } else if (r < 0) {
    return 0.0;
  }

  vector<vector<double>>& P = *P_ptr;
  if (P[r][n] == -1.0) {
    P[r][n] =
        HouseMajorityHelper(prob, r - 1, n - 1, P_ptr) * prob[n - 1] +
        HouseMajorityHelper(prob, r, n - 1, P_ptr) * (1.0 - prob[n - 1]);
  }
  return P[r][n];
}
// @exclude

void PrintVector(vector<double> prob) {
  sort(prob.begin(), prob.end());
  for (int i = 0; i < prob.size(); ++i) {
    cout << i << ":" << static_cast<int>(100 * prob[i])
         << (((i + 1) % 10 == 0) ? "\n" : " ");
  }
  cout << endl;
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  int n = 435;
  vector<double> prob;
  prob.emplace_back(0.0);
  for (int i = 0; i < n; ++i) {
    uniform_real_distribution<double> dis(0, 1);
    prob.emplace_back(dis(gen));
  }
  PrintVector(prob);
  double ans = HouseMajority(prob, n);
  assert(0.0 <= ans && ans <= 1.0);
  cout << ans << endl;
  return 0;
}
